# 1. Tsfresh: Univariate Feature Engineering with Robot Failure Dataset (LP1)

This basic example shows how to use [tsfresh](https://tsfresh.readthedocs.io/) to extract useful features from univaraite timeseries and use them to improve classification performance.

We use the gunpoint dataset as an example.

In [6]:
import pandas as pd
from tsfresh import extract_features, extract_relevant_features, select_features
from tsfresh.utilities.dataframe_functions import impute
from tsfresh.feature_extraction import ComprehensiveFCParameters

#load dataset
lp1 = pd.read_csv(r'C:\Users\seren\Documents\RR Lab Documents\lp1_fulldata.csv')

y = lp1['class']
lp1 = lp1.drop(['class'], axis=1)
lp1 = lp1.iloc[:, :3] #univariate

In [7]:
lp1.head()

id  time  var1
0   1     1    -1
1   1     2     0
2   1     3    -1
3   1     4    -1
4   1     5    -1

In [8]:
extraction_settings = ComprehensiveFCParameters()

X = extract_features(lp1, column_id='id', column_sort='time',
                     default_fc_parameters=extraction_settings,
                     # we impute = remove all NaN features automatically
                     impute_function=impute)

Feature Extraction: 100%|██████████████████████████████████████████████████████████████| 18/18 [00:07<00:00,  2.35it/s]


In [9]:
print('Extracted Feature set shape: ', X.shape)
X['class'] = y
print('Extracted Feature set shape with class labels: ', X.shape)

Extracted Feature set shape:  (88, 779)
Extracted Feature set shape with class labels:  (88, 780)


# 2. Tsfresh: Multivariate Feature Engineering with Robot Failure Dataset (LP1)

This basic example shows how to use [tsfresh](https://tsfresh.readthedocs.io/) to extract useful features from multiple timeseries and use them to improve classification performance.

We use the robot failure data set as an example.

In [10]:
import pandas as pd
from tsfresh import extract_features, extract_relevant_features, select_features
from tsfresh.utilities.dataframe_functions import impute
from tsfresh.feature_extraction import ComprehensiveFCParameters

## Extract Features

In [11]:
#load dataset
lp1 = pd.read_csv(r'C:\Users\seren\Documents\RR Lab Documents\lp1_fulldata.csv')

y = lp1['class']
lp1 = lp1.drop(['class'], axis=1)

In [12]:
lp1.head()

id  time  var1  var2  var3  var4  var5  var6
0   1     1    -1     0    57    -5    -3     0
1   1     2     0    -3    63    -1     0     0
2   1     3    -1     1    51    -4    -1    -1
3   1     4    -1    -2    68    -2    -2     0
4   1     5    -1    -1    65    -6     1     0

In [13]:
extraction_settings = ComprehensiveFCParameters()

X = extract_features(lp1, column_id='id', column_sort='time',
                     default_fc_parameters=extraction_settings,
                     # we impute = remove all NaN features automatically
                     impute_function=impute)

Feature Extraction: 100%|██████████████████████████████████████████████████████████████| 20/20 [00:23<00:00,  1.15s/it]


It should be noted that given that there are 6 variables, this results in the 779*6 columns to be created. 

In [14]:
print('Extracted Feature set shape: ', X.shape)
print('Number of features extracted per sensor: ', X.shape[1]/6)
X = X.reset_index(drop = True)
X['class'] = y[::15].reset_index(drop = True)
print('Extracted Feature set shape with class labels: ', X.shape)

Extracted Feature set shape:  (88, 4674)
Number of features extracted per sensor:  779.0
Extracted Feature set shape with class labels:  (88, 4675)


In [15]:
#check for nan
is_NaN = X.isnull()
row_has_NaN = is_NaN.any(axis=1)
rows_with_NaN = X[row_has_NaN]

rows_with_NaN

Empty DataFrame
Columns: [var4__variance_larger_than_standard_deviation, var4__has_duplicate_max, var4__has_duplicate_min, var4__has_duplicate, var4__sum_values, var4__abs_energy, var4__mean_abs_change, var4__mean_change, var4__mean_second_derivative_central, var4__median, var4__mean, var4__length, var4__standard_deviation, var4__variation_coefficient, var4__variance, var4__skewness, var4__kurtosis, var4__absolute_sum_of_changes, var4__longest_strike_below_mean, var4__longest_strike_above_mean, var4__count_above_mean, var4__count_below_mean, var4__last_location_of_maximum, var4__first_location_of_maximum, var4__last_location_of_minimum, var4__first_location_of_minimum, var4__percentage_of_reoccurring_values_to_all_values, var4__percentage_of_reoccurring_datapoints_to_all_datapoints, var4__sum_of_reoccurring_values, var4__sum_of_reoccurring_data_points, var4__ratio_value_number_to_time_series_length, var4__sample_entropy, var4__maximum, var4__minimum, var4__benford_correlation, var4__time_reversal_asymmetry_statistic__lag_1, var4__time_reversal_asymmetry_statistic__lag_2, var4__time_reversal_asymmetry_statistic__lag_3, var4__c3__lag_1, var4__c3__lag_2, var4__c3__lag_3, var4__cid_ce__normalize_True, var4__cid_ce__normalize_False, var4__symmetry_looking__r_0.0, var4__symmetry_looking__r_0.05, var4__symmetry_looking__r_0.1, var4__symmetry_looking__r_0.15000000000000002, var4__symmetry_looking__r_0.2, var4__symmetry_looking__r_0.25, var4__symmetry_looking__r_0.30000000000000004, var4__symmetry_looking__r_0.35000000000000003, var4__symmetry_looking__r_0.4, var4__symmetry_looking__r_0.45, var4__symmetry_looking__r_0.5, var4__symmetry_looking__r_0.55, var4__symmetry_looking__r_0.6000000000000001, var4__symmetry_looking__r_0.65, var4__symmetry_looking__r_0.7000000000000001, var4__symmetry_looking__r_0.75, var4__symmetry_looking__r_0.8, var4__symmetry_looking__r_0.8500000000000001, var4__symmetry_looking__r_0.9, var4__symmetry_looking__r_0.9500000000000001, var4__large_standard_deviation__r_0.05, var4__large_standard_deviation__r_0.1, var4__large_standard_deviation__r_0.15000000000000002, var4__large_standard_deviation__r_0.2, var4__large_standard_deviation__r_0.25, var4__large_standard_deviation__r_0.30000000000000004, var4__large_standard_deviation__r_0.35000000000000003, var4__large_standard_deviation__r_0.4, var4__large_standard_deviation__r_0.45, var4__large_standard_deviation__r_0.5, var4__large_standard_deviation__r_0.55, var4__large_standard_deviation__r_0.6000000000000001, var4__large_standard_deviation__r_0.65, var4__large_standard_deviation__r_0.7000000000000001, var4__large_standard_deviation__r_0.75, var4__large_standard_deviation__r_0.8, var4__large_standard_deviation__r_0.8500000000000001, var4__large_standard_deviation__r_0.9, var4__large_standard_deviation__r_0.9500000000000001, var4__quantile__q_0.1, var4__quantile__q_0.2, var4__quantile__q_0.3, var4__quantile__q_0.4, var4__quantile__q_0.6, var4__quantile__q_0.7, var4__quantile__q_0.8, var4__quantile__q_0.9, var4__autocorrelation__lag_0, var4__autocorrelation__lag_1, var4__autocorrelation__lag_2, var4__autocorrelation__lag_3, var4__autocorrelation__lag_4, var4__autocorrelation__lag_5, var4__autocorrelation__lag_6, var4__autocorrelation__lag_7, var4__autocorrelation__lag_8, var4__autocorrelation__lag_9, ...]
Index: []

[0 rows x 4675 columns]

In [16]:
X.to_csv(r'C:\Users\seren\Documents\oct\tsfresh_multi.csv', index = False)